File: go/rerout/project.go
// Package rerout — Project namespace.
//
// Reached via Client.Project(). Wraps the /v1/projects/me/* surface:
// aggregate stats and current-project info.

package rerout

import (
	"context"
	"net/url"
)

// ProjectNS is the project-operations namespace. Construct via
// Client.Project(). The trailing "NS" disambiguates from the Project response
// type.
type ProjectNS struct {
	client *Client
}

// Stats returns aggregate analytics across every link in the project.
// Defaults to 30 days when days <= 0.
func (p *ProjectNS) Stats(ctx context.Context, days int) (*ProjectStats, error) {
	if days <= 0 {
		days = 30
	}
	q := url.Values{}
	q.Set("days", intToString(days))
	var out ProjectStats
	err := p.client.do(ctx, requestOptions{
		method: "GET",
		path:   "/v1/projects/me/stats",
		query:  q,
	}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// Me returns info about the project that owns the current API key.
func (p *ProjectNS) Me(ctx context.Context) (*Project, error) {
	var out Project
	err := p.client.do(ctx, requestOptions{
		method: "GET",
		path:   "/v1/projects/me",
	}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

File: ruby/lib/rerout/project.rb
# frozen_string_literal: true

module Rerout
  module Resources
    # Project-level operations namespace.
    class Project
      # @param client [Rerout::Client]
      def initialize(client)
        @client = client
      end

      # Aggregate stats across every link in the project. Defaults to 30 days.
      #
      # @param days [Integer]
      # @return [Rerout::Models::ProjectStats]
      def stats(days: 30)
        response = @client.request(
          method: :get,
          path: '/v1/projects/me/stats',
          query: { 'days' => days }
        )
        Models::ProjectStats.from_hash(response)
      end

      # Info about the project that owns the current API key.
      #
      # @return [Rerout::Models::Project]
      def me
        response = @client.request(method: :get, path: '/v1/projects/me')
        Models::Project.from_hash(response)
      end
    end
  end
end
